wallet-new/src/Cardano/Wallet/Kernel/DB/HdWallet/Derivation.hs

```diff
@@ -0,0 +1,31 @@
+module Cardano.Wallet.Kernel.DB.HdWallet.Derivation (
+      deriveIndex
+    , HardeningMode(..)
+    ) where
+
+import           Universum
+
+import           Cardano.Crypto.Wallet.Types (DerivationIndex)
+
+data HardeningMode = SoftDerivation
+                   -- ^ Generates indexes in the range (0, maxBound @Word32)
+                   | HardDerivation
+                   -- ^ Generates indexes in the range (0x8000000, maxBound @Word32)
+
+-- | Derives a new '
+deriveIndex :: Monad m
+            => ((DerivationIndex, DerivationIndex) -> m DerivationIndex)
+            -- ^ A monadic computation which can pick a 'DerivationIndex' out
+            -- of a range.
+            -> (DerivationIndex -> a)
+            -- How to build the final type out of the picked 'DerivationIndex'.
+            -> HardeningMode
+            -- ^ How we want to derive this index (@soft@ vs @hard@)
+            -> m a
+deriveIndex pickRange mkA hardeningMode =
+    let range = case hardeningMode of
+                     SoftDerivation -> (0, maxBound)
+                     HardDerivation -> (0x8000000, maxBound)
```

Comment:
Plan of action, to avoid the magic number here:

- Let's use `firstHardened` from `Pos.Crypto.HD`
- Have `cardano-crypto` to export `0x8000000` as top-level constant
- Let's have `firstHardened` to alias to this new exported value.